rfcs/0016-error-handling.md

```diff
@@ -0,0 +1,379 @@
+
+This RFC aims for a clear guideline on how to properly utilize ABI errors, `panic` expression in particular.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Throughout this RFC, we use the following terms:
+- _end-developer_ for application developers, who write they own code over which they have full control, and reuse Sway libraries, over whose implementation and behavior they have no control.
+- _reusable library_ for any dependency, that the end-developer has no control over, i.e., cannot opt in or out of a certain behavior of individual functions/methods. This includes `std`, libraries provided by Fuel, like, e.g., `sway_libs`, or any third-party libraries.
+- _application_ for any end-developer code. This will usually mean contracts, scripts, or predicates, but can also mean application specific local libraries, that the end-developer has a full control of.
+
+When it comes to error handling, end-developers cannot influence implementation and behavior of _reusable libraries_ they use as dependencies. That's why we expect _reusable libraries_ to follow strict guidelines, to fullfil the expectations listed in the below chapter.
+
+## Error handling in reusable libraries
+
+[error-handling-in-reusable-libraries]: #error-handling-in-reusable-libraries
+
+When using _reusable libraries_ the end-developers will expect the following:
+1. In case of irrecoverable errors, the error location will always be in the library code that has caused the issue.
+1. In case of irrecoverable errors, there will be no on-chain cost for error handling by default, unless the developer opts-in for a cost via dedicated API.
+1. Error type enums will have the prefix `Error`. E.g., `AccessError`.
+1. Error messages will be helpful and follow a consistent pattern and wording.
+
+The above expectations imply the following rules for implementing _reusable libraries_:
+1. _Reusable libraries_ will never use helper functions like `assert`, `revert`, or `require`. **Because, in the case of using those helpers, the error location will be in the helper and not in the library code.** Instead, they will use the `panic` expression directly, ensuring the error location being in the library code.
+1. _Reusable libraries_ will exclusively use strings for formulating error messages in the default API. Using enums would result in increase of bytecode size that cannot be opted in or out by end-developers. Providing additional information via non-unit enum variants might be useful in certain cases. For such cases, a dedicated opt-in API should be provided. An example would be `Result::unwrap` and `Result::expect` methods. The former, considered the default, will `panic` with zero-cost and useful common error message, while the latter will provide rich information, but with on-chain overhead. Important is that the developer can opt-in and accept that cost and is not getting it as default.
```

Comment:
Why is it the case that Enums imply a higher cost? Is it because of the encoding logic? You'd think that would be minimal.

I don't really like the idea of libraries being unable to share common errors and their structure where appropriate and having to rely on pure strings.